src/SpecsParser.sol

```diff
@@ -238,34 +248,89 @@ library ERC4337SpecsParser {
+     * @param debugTrace The debug trace to validate
+     * @param entities  The entities of the userOp
+     * @param userOpDetails The UserOperationDetails to validate
      */
+    // solhint-disable-next-line code-complexity
     function validateBannedStorageLocations(
-        VmSafe.AccountAccess memory currentAccess,
-        Entities memory entities
+        VmSafe.DebugStep[] memory debugTrace,
+        Entities memory entities,
+        UserOperationDetails memory userOpDetails
     )
         internal
     {
-        // Loop over the storage accesses
-        for (uint256 j; j < currentAccess.storageAccesses.length; j++) {
-            // Get current storage access and account
-            VmSafe.StorageAccess memory currentStorageAccess = currentAccess.storageAccesses[j];
-            address currentAccessAccount = currentStorageAccess.account;
-
-            // Allow storage accesses from the sender or allowed entities
-            if (!isEntityAndStaked(entities, currentAccessAccount)) {
-                bytes32 currentSlot = currentStorageAccess.slot;
-                bool isAssociated = isAssociatedStorage(currentSlot, currentAccessAccount, entities);
-                if (!isAssociated) {
+        // Loop over the debug steps to validate the storage accesses
+        for (uint256 i; i < debugTrace.length; i++) {
+            // Get the current debug step
+            VmSafe.DebugStep memory currentStep = debugTrace[i];
+            // Skip if the opcode is not SLOAD or SSTORE, TLOAD or TSTORE
+            if (
+                currentStep.opcode != 0x54 // SLOAD
+                    && currentStep.opcode != 0x55 // SSTORE
+                    && currentStep.opcode != 0x5C // TLOAD
+                    && currentStep.opcode != 0x5D // TSTORE
+            ) {
+                continue;
+            }
+            // Init current access account
+            address currentAccessAccount = currentStep.contractAddr;
+            // Init current access slot
+            bytes32 currentSlot = bytes32(uint256(currentStep.stack[0]));
+
+            /// Access to associated storage of the account in an external (non-entity) contract
+            // is allowed if either
+            if (!isEntity(entities, currentAccessAccount)) {
+                // [STO-021] The account already exists.
+                bool accountAlreadyExists = entities.account.code.length != 0
+                    || (
+                        currentAccessAccount == userOpDetails.entryPoint
+                            && entities.account != address(0)
+                    );
+                // [STO-022] There is an initCode and the factory contract is staked.
+                bool isFactoryStaked = entities.isFactoryStaked;
+                if (
+                    !(
+                        isAssociatedStorage(currentSlot, currentAccessAccount, entities)
+                            && (accountAlreadyExists || isFactoryStaked)
```

Comment:
if factory is staked then only storage associated with the factory is allowed right? eg wouldnt this not revert if someone tries to access associated storage with account that does not exist but factory is staked?

Reply:
true, mixed up currentAccessAccount with "account" in the specs

Reply:
fixed in [5165b53]. Added a new version of isAssociatedStorage that checks for association with a specific entity, the old version checked for association with any entity